README.md
# AgileCLU #

AgileCLU is a command line tool implementation and Python programming library for Limelight Networks Agile Storage cloud platform.  It leverages Agile's JSON-RPC APIs and HTTP ingest and egress capabilities in an easy to use way.  To use these tools, you must have:

* An account on Limelight Network's Agile Storage cloud platform. (http://www.limelightnetworks.com)


## Communication ##

Feel free to send any questions, comments, or patches to my Github page (you'll need to join to send a message): 
https://github.com/wylieswanson/AgileCLU


## Basic Installation ##
If you already have Python and [Python Package Index](http://pypi.python.org/pypi/setuptools) (PyPI) installed on your machine, the installation of AgileCLU is simple and straightfoward.  Simply execute one of the following commands (sudo is usually required on Linux):

	easy_install AgileCLU

or,

	pip install AgileCLU

If the above method worked for you, you can skip the operating system-specific installation sections and move to Configuration, as you have now completed the installation of AgileCLU.  If not, consult the relevant operating system-specific sections in the Advanced Configuration sections.

## Upgrading ##

If you are upgrading from a release prior to 0.3.1, you may need to manually delete the files from your Python installation (egg and easy-install.pth) prior to invoking easy_install or pip.  For future upgrades, can force to latest version with:

	easy_install -U AgileCLU

# Configuration #

After installing AgileCLU, run use the profile tool to generate the proper output to place in an /etc/agile/agile.conf configuration file.  You can create as many configuration profiles as you like, specifying to use them over the default agile.conf profile by specifying the -l option on any given command.  If you are using Windows, you need to include the .py extension, substituting agileprofile with agileprofile.py.

	agileprofile

Example output:

	agileprofile (AgileCLU 0.3.6)
	
	This tool generates Agile Storage profile text to be pasted into AgileCLU profile configuration
	files (eg. /etc/agile/agile.conf is the default).  The information should have been provided to
	you by Limelight Networks, often via a "welcome letter".  If you have questions regarding the
	account information, please contact support@llnw.com.
	
	Enter Agile username: testcompany
	Enter Agile password: 
	Re-enter Agile password: 
	Enter your egress hostname: global.mt.lldns.net
	Enter your egress base path: /testcompany
	Enter your ingest hostname: api.agile.lldns.net
	
	PASTE THE FOLLOWING IN YOUR PROFILE CONFIGURATION FILE 
	
	[Identity]
	username = testcompany
	password = A4UsWnRpKOdNy0HNWDHY+Q==
	
	[Egress]
	protocol = http
	hostname = global.mt.lldns.net
	basepath = /testcompany
	
	[Ingest]
	protocol = https
	hostname = api.agile.lldns.net
	
	[Logging]
	enabled = no
	logfile = /var/log/agileclu.log
	level = info


# AgileCLU from Command Line #

The commands that are currently available are:

*agileprofile* - Generate a profile based on account credentials and ingest/egress information

*agilefetch* - Automatically download a file from any URL and place it in your storage in a specified directory

*agilemkdir* - Make a directory

*agilerm* - Remove a file

*agilels* - List a directory

*agilepost* - Upload a file

NOTE: For Windows, add a ".py" extension to the above commands.





# Advanced Installation #

The advanced installation covers installing prerequisites, like Python and Python Setuptools.  Specific Python libraries will be installed automaticaly when you run easy_install.  If you already have Python and Easysetup installed, you do not need to use the following directions.


## Installation:Linux ##

On most Linux distributions, Python is already installed, you only need to install PyPI.  For Debian, Ubuntu and other distributions using APT, install PyPI with the following:

	sudo apt-get install python-pip

If you are running another distribution, consult the [Python setuptools](http://pypi.python.org/pypi/setuptools) documentation.  After you complete this step, complete Basic Installation and move on to Configuration.


## Installation:Mac OSX ##

Python is already installed by default on modern OS X.

## Installation:Windows 32-bit and 64-bit ##

The Windows 32-bit and 64-bit Installation section covers Windows environment variables, along with Python and Python Setuptools.

### Windows Python ###

Python must be installed on the machine.  You can download from http://www.python.org/getit/ or, specifically, for Windows 32 and 64-bit:

* Python 2.7.3 Windows Installer (Windows binary - does not include source)
	* http://www.python.org/ftp/python/2.7.3/python-2.7.3.msi

* Python 2.7.3 Windows X86-64 Installer (Windows AMD64 / Intel 64 / X86-64 bainry - does not include source)
	* http://www.python.org/ftp/python/2.7.3/python-2.7.3.amd64.msi

### Windows Environment Variables ###

Once Python has been installed, you will want to add setuptools, the mainstream package manager for Python, also known as PyPI.

Next, set the system's PATH variable to include directories that include Python components and packages we'll add later.  To do this:

* Click the bottom left Windows icon
* In the search field, type 'system'
* In the Control Panel section of the search results, select "Edit system environment variables"
* Select "Environment Variables"
* In the "System variables" section, scroll down to Path and click "Edit...", and then append the below text to the "Variable Value" field, then select OK.

> ;C:\Python27;C:\Python27\Lib\site-packages;C:\Python27\Scripts;


### Windows Python Setuptools ###

* For 32-bit Windows
	* Install setuptools using the provided .exe installer.
		* http://pypi.python.org/packages/2.7/s/setuptools/setuptools-0.6c11.win32-py2.7.exe

* For 64-bit Windows
	* Download ez_setup.py and run it; it will download the appropriate .egg file and install it for you. (Currently, the provided .exe installer does not support 64-bit versions of Python for Windows, due to a distutils installer compatibility issue.
		* http://peak.telecommunity.com/dist/ez_setup.py
		* Run "ez_setup.py"

At this point, you can return to the basic installation method (easy_install) at the top of this document.  Note that you will need to place the output of agileprofile in C:\etc\agile\agile.conf, or alternate profiles C:\etc\agile\profileconf (to be used by the -l command line option).


# Libraries used by AgileCLU #

This package leverages the following Python libraries:

* poster by Chris AtLee - used for streaming ingest (http://atlee.ca/software/poster/)
* progressbar by Nilton Volpato - used for console ingest progress bar (http://code.google.com/p/python-progressbar/)
* pydes by Todd Whiteman - used as part of the password encryption scheme for config files (http://twhiteman.netfirms.com/des.html)
* jsonrpclib by John Marshall - an implementation of the JSON-RPC specification (https://github.com/joshmarshall/jsonrpclib)


# Agile Storage Locations #

As of October 2012, the Agile Storage Cloud has storage capacity in 34 geographies around the world.

![Agile Storage Locations](https://raw.github.com/wylieswanson/AgileCLU/master/agile_locations_oct_2012.jpg)


CHANGES.md
CHANGES
=======

* 0.3.1 - configuration file format change
* 0.3.0 - major overhaul, re-read README!  


src/AgileCLU.py
#!/usr/bin/python
# Filename: AgileCLU.py
# coding: utf-8
#
# Copyright (C) 2010-2011, Wylie Swanson
#
# This Program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License V3, as published by
# the Free Software Foundation.
#
# This Program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.

from jsonrpc import ServiceProxy
import ConfigParser, sys, os.path, logging
import poster 
from urllib2 import Request, urlopen, URLError, HTTPError

logger = logging.getLogger('AgileCLU')
hdlr = logging.FileHandler( '/var/log/agileclu.log' )
formatter = logging.Formatter('%(asctime)s %(levelname)s %(message)s')
hdlr.setFormatter(formatter)
logger.addHandler(hdlr)
logger.setLevel(logging.INFO)


cfg = ConfigParser.ConfigParser() 
version = '0.2'

class	AgileCLU:

        def     __init__(self, username='agile'):

		# Load configuration variables

		if os.path.exists('/etc/agile/'+username+'.conf'): cfg.read('/etc/agile/'+username+'.conf')
		else:
			print "Alternate login identity (%s) configuration does not exist.  Exiting." % username
			logger.critical( "configuration /etc/agile/"+username+".conf does not exist" )
			sys.exit(1)

		self.uid = cfg.get("Identity", "username")
		upw = cfg.get("Identity", "password")
		self.apiurl = cfg.get("Ingest", "apiurl")
		self.posturl = cfg.get("Ingest", "posturl")
		self.mapperurl = cfg.get("Egress", "mapperurl")
		self.cacheurl = cfg.get("Egress", "mapperurl")

		# connect to API, authenticate, and get a token

		self.api = ServiceProxy( self.apiurl )
		self.token, self.user = self.api.login( self.uid, upw )
		if self.token is None:
			print "Autentication for '"+str(self.uid)+"' failed. Please check credentials.  Exiting."
			logger.critical( self.uid+" login failed - check account credentials" )
			sys.exit(1)

		logger.info( self.uid+" "+self.token+", login "+str(self.user)+" tokenized" )
	
	def	apiurlstr(self):
		logger.info( self.uid+" "+self.token+", apiurl="+self.apiurl )
		return self.apiurl

	def	posturlstr(self):
		logger.info( self.uid+" "+self.token+", posturl="+self.posturl )
		return self.posturl

	def	mapperurlstr(self):
		logger.info( self.uid+" "+self.token+", mapperurl="+self.mapperurl )
		return self.mapperurl

	def	cacheurlstr(self):
		logger.info( self.uid+" "+self.token+", cacheurl="+self.cacheurl )
		return self.cacheurl

	def	tokenstr(self):
		logger.info( self.uid+" "+self.token+", self.token" )
		return self.token

	def	stat(self, path):
		r = self.api.stat( self.token, path )
		logger.info( self.uid+" "+self.token+", stat "+path+" = "+str(r) )
		return r

	def	logout(self):
		r = self.api.logout( self.token )
		logger.info( self.uid+" "+self.token+", logout = "+str(r)+", detokenized" )
		return r

	def	noop(self):
		r = self.api.noop( self.token )
		logger.info( self.uid+" "+self.token+", noop = "+str(r) )
		return r

	def	listDir(self, path, pageSize=10000, cookie=0, stat=True ):
		r = self.api.listDir( self.token, path, pageSize, cookie, stat )
		logger.info( self.uid+" "+self.token+", listDir "+path+" pageSize "+str(pageSize)+" cookie "+str(cookie)+" stat "+str(stat) )
		# +" = "+str(r) )
		return r

	def	listFile(self, path, pageSize=10000, cookie=0, stat=True ):
		r = self.api.listFile( self.token, path, pageSize, cookie, stat )
		logger.info( self.uid+" "+self.token+", listFile "+path+" pageSize "+str(pageSize)+" cookie "+str(cookie)+" stat "+str(stat) )
		# +" = "+str(r) )
		return r

	def	makeDir(self, path):
		r = self.api.makeDir( self.token, path)
		logger.info( self.uid+" "+self.token+", makeDir "+path+" = "+str(r) )
		return r

	def	makeDir2(self, path):
		r = self.api.makeDir2( self.token, path)
		logger.info( self.uid+" "+self.token+", makeDir2 "+path+" = "+str(r) )
		return r

	def	deleteFile(self, path):
		r = self.api.deleteFile( self.token, path)
		logger.info( self.uid+" "+self.token+", deleteFile "+path+" = "+str(r) )
		return r

	def	rm(self, path):
		if (self.fexists(path)):
			r = self.deleteFile(path)
			if (r == 0): 
				logger.info( self.uid+" "+self.token+", rm "+path+" succeeded" )
				return True
			else: 
				logger.warning( self.uid+" "+self.token+", rm "+path+" failed" )
				return False
		else:
			logger.warning( self.uid+" "+self.token+", rm "+path+" skipped nonexistent file" )

	def	deleteDir(self, path):
		r = self.api.deleteDir( self.token, path)
		logger.info( self.uid+" "+self.token+", deleteDir "+path+" = " + str(r) )
		return r

	def	deleteObject(self, path):
		r = self.api.deleteObject( self.token, path)
		logger.info( self.uid+" "+self.token+", deleteObject "+path+" = " + str(r) )
		return r

	def	rename(self, path, newpath):
		r = self.api.rename( self.token, path, newpath)
		logger.info( self.uid+" "+self.token+", rename "+path+" to "+newpath+" = " + str(r) )
		return r
	
	def	copyFile(self, path, newpath):
		r = self.api.copyFile( self.token, path, newpath)
		logger.info( self.uid+" "+self.token+", copyFile "+path+" to "+newpath+" = " + str(r) )
		return r

	def	registerCallback( self, uri, flags=0, threshold=0 ):
		r = self.api.registerCallback( self.token, uri, flags, threshold )
		logger.info( self.uid+" "+self.token+", registerCallback "+uri+" flags "+str(flags)+" threshold "+str(threshold)+" = " + str(r) )
		return r

	def	listCallback( self ):
		r = self.api.listCallback( self.token )
		logger.info( self.uid+" "+self.token+", listCallback" )
		return r

	def	fetchFileHTTP( self, path, uri, username=None, password=None, auth=None, callbackid=0, priority=0, flags=0, expose_egress='COMPLETE'):
		r = self.api.fetchFileHTTP( self.token, path, uri, username, password, auth, callbackid, priority, flags, expose_egress)
		logger.info( self.uid+" "+self.token+", fetchFileHTTP path "+path+" uri "+uri+" username "+str(username)+" auth "+str(auth)+" callbackid "+str(callbackid)+" priority "+str(priority)+" flags "+str(flags)+" expose_egress "+expose_egress+" = " + str(r) )
		return r

	def	fetchFileFTP(self, path, hostname, filename, username=None, password=None, port=21, passive=True, callbackid=0, priority=0, flags=0, expose_egress='COMPLETE'):
		r = self.api.fetchFileFTP( self.token, path, hostname, filename, username, password, port, passive, callbackid, priority, flags, expose_egress)
		logger.info( self.uid+" "+self.token+", fetchFileFTP path "+path+" hostname "+hostname+" filename "+filename+" username "+str(username)+" port "+str(port)+" passive "+str(passive)+" callbackid "+str(callbackid)+" priority "+str(priority)+" flags "+str(flags)+" expose_egress "+expose_egress+" = " + str(r) )
		return r

	def	fexists( self, path ):
		r = self.stat( path )
		logger.info( self.uid+" "+self.token+", fexists "+path+" = "+str(r) )
		if ((r['code'] == 0) and (r['type'] == 2)): return True
		else: return False

	def	dexists( self, path ):
		r = self.stat( path )
		logger.info( self.uid+" "+self.token+", dexists "+path+" = "+str(r) )
		if ((r['code'] == 0) and (r['type'] == 1)): return True
		else: return False

	def	exists(self, path):
		logger.info( self.uid+" "+self.token+", exists "+path )
		# +" = "+str(r) )
		if (self.fexists(path) or self.dexists(path)): return True
		else: return False

	def	mkdir(self, path, recursive = False):
		logger.info( self.uid+" "+self.token+", mkdir "+path )
		# +" = "+str(r) )
		if (not self.dexists(path)):
			if (recursive):
				r = self.makeDir2( path )
			else:
				r = self.makeDir( path )
			if (r == 0): 
				return True
			else: 
				return False
		else:	
			return False

	def	read(self, path):
		logger.info( self.uid+" "+self.token+", read "+self.mapper.url+urllib2.quote(path) )
		if (self.fexists(path)):
			response = urllib2.urlopen( self.mapperurl+urllib2.quote(path) )
			return response.read()
		else:
			return False

	def	post(self, source, destination, rename=None, mimetype='auto', mtime=None, egress_policy='COMPLETE', mkdir=False, callback=None):
		logger.info( self.uid+" "+self.token+", post "+source+" "+destination+", rename="+str(rename)+", mimetype="+str(mimetype)+", mtime="+str(mtime)+", egress="+str(egress_policy)+", mkdir="+str(mkdir))
		if (not os.path.isfile(source)): logger.info( "local("+source+") does not exist") ; return False 
		if (not self.dexists(destination)): logger.info( "remote("+destination+") does not exist") ; return False

		source_path = os.path.dirname(source) ; source_name = os.path.basename(source)
	
		poster.streaminghttp.register_openers()
	
		if callback<>None:
			datagen, headers = poster.encode.multipart_encode( {
				"uploadFile": open(source, "rb"),
				"directory": destination,
				"basename": source_name,
				"expose_egress": egress_policy
				}, cb=callback)
		else:
			datagen, headers = poster.encode.multipart_encode( {
				"uploadFile": open(source, "rb"),
				"directory": destination,
				"basename": source_name,
				"expose_egress": egress_policy
				} )

		request = Request(self.posturl, datagen, headers)
		request.add_header("X-Agile-Authorization", self.token)
		request.add_header("X-Content-Type", mimetype )

		try: result = urlopen(request).read()
		except HTTPError, e: logger.info( 'HTTP Error: '+str(e.code) ) ; return False
		except URLError, e: logger.info( 'URL Error: '+str(e.reason) ) ; return False

		return True

# End of agile.py


AgileCLU/__init__.py
import jsonrpclib
import ConfigParser, sys, os.path, logging
import poster 
import pyDes, hashlib, base64
from urllib2 import Request, urlopen, URLError, HTTPError
import progressbar

class NullHandler(logging.Handler):
    def emit(self, record):
        pass

logger = logging.getLogger('AgileCLU')
logger.addHandler(NullHandler())
cfg = ConfigParser.ConfigParser() 

def epwbasekey( username, proto, hostname, basepath ):
   m = hashlib.md5()
   m.update( username )
   m.update( proto )
   m.update( hostname )
   m.update( basepath )
   key = base64.b64encode(hashlib.sha256( m.digest() ).digest())
   return key[0:24]

def e_pw_hash( str, username, proto, hostname, basepath ):
   return base64.b64encode(pyDes.triple_des(epwbasekey(username,proto,hostname,basepath)).encrypt(str, padmode=2))

def e_pw_dehash( str, username, proto, hostname, basepath ):
	basekey=epwbasekey(username,proto,hostname,basepath)
	try:
		b64decode=base64.b64decode(str)
	except TypeError, ValueError:
		b64decode="12345678"
	try:
		if (len(b64decode) % 8 == 0):
			dehash = pyDes.triple_des(basekey).decrypt(b64decode, padmode=2)
		else:
			print "Password is not valid. Verify profile was built with \"agileprofile\" command."
			sys.exit(1)
	except TypeError, ValueError:
		dehash = "87654321"
	if dehash is '': dehash='87654321' 
	return dehash

class	AgileCLU:
	__module__ = "AgileCLU"
	__version__ = "0.3.12"

	def     __init__(self, profile='agile'):

		# Load configuration variables

		if os.path.exists('/etc/agile/'+profile+'.conf'): cfg.read('/etc/agile/'+profile+'.conf')
		else:
			print "Profile (%s) does not exist.  Create profile with \"agileprofile\" command." % profile
			# logger.critical( "configuration /etc/agile/"+profile+".conf does not exist" )
			sys.exit(1)

		# minor check to look for compliant configuration file
		if not cfg.has_option("Egress","protocol"):
			print "Profile (%s) is not valid!  Create profile with \"agileprofile\" command." % profile
			sys.exit(1)


		self.uid = cfg.get("Identity", "username")

		self.egress_protocol = cfg.get("Egress", "protocol")
		self.egress_hostname = cfg.get("Egress", "hostname")
		self.egress_basepath = cfg.get("Egress", "basepath")
		
		self.ingest_protocol = cfg.get("Ingest", "protocol")
		self.ingest_hostname = cfg.get("Ingest", "hostname")

		self.mapperurl = self.egress_protocol + "://" + self.egress_hostname + self.egress_basepath
	
		self.apiurl = self.ingest_protocol + "://" + self.ingest_hostname + "/jsonrpc"
		self.posturl = self.ingest_protocol + "://" + self.ingest_hostname + "/post/file"
		self.postmultiurl = self.ingest_protocol + ":8080//" + self.ingest_hostname + "/multipart"

		self.pbar = None
		self.pbarfname = None

		upw = e_pw_dehash( 
			cfg.get("Identity", "password"), 
			self.uid, 
			self.egress_protocol, 
			self.egress_hostname, 
			self.egress_basepath )

		if upw is "87654321":
			print "Password is not valid!  Verify profile was built with \"agileprofile\" command."
			sys.exit(1)

		# initialize the logger for session
		if cfg.getboolean("Logging", "enabled" ):
			hdlr = logging.FileHandler( '/var/log/agileclu.log' )
			formatter = logging.Formatter('%(asctime)s %(levelname)s %(message)s')
			hdlr.setFormatter(formatter)
			logger.addHandler(hdlr)
			logger.setLevel(logging.INFO)

		# connect to API, authenticate, and get a token
		self.api = jsonrpclib.Server( self.apiurl )
		self.token, self.user = self.api.login( self.uid, upw )
		if self.token is None:
			print "Authentication for '"+str(self.uid)+"' failed!  It is recommended to generate a new profile with \"agileprofile\" command."
			logger.critical( self.uid+" login failed - check account credentials" )
			sys.exit(1)

		logger.info( self.uid+" "+self.token+", login "+str(self.user)+" tokenized" )
	
	def	apiurlstr(self):
		logger.info( self.uid+" "+self.token+", apiurl="+self.apiurl )
		return self.apiurl

	def	posturlstr(self):
		logger.info( self.uid+" "+self.token+", posturl="+self.posturl )
		return self.posturl

	def	mapperurlstr(self):
		logger.info( self.uid+" "+self.token+", mapperurl="+self.mapperurl )
		return self.mapperurl

	def	cacheurlstr(self):
		logger.info( self.uid+" "+self.token+", cacheurl="+self.cacheurl )
		return self.cacheurl

	def	postmultiurlstr(self):
		logger.info( self.uid+" "+self.token+", postmultiurl="+self.postmultiurl )
		return self.postmultiurl

	def	tokenstr(self):
		logger.info( self.uid+" "+self.token+", self.token" )
		return self.token

	def	stat(self, path):
		r = self.api.stat( self.token, path )
		logger.info( self.uid+" "+self.token+", stat "+path+" = "+str(r) )
		return r

	def	logout(self):
		r = self.api.logout( self.token )
		logger.info( self.uid+" "+self.token+", logout = "+str(r)+", detokenized" )
		return r

	def	noop(self):
		r = self.api.noop( self.token )
		logger.info( self.uid+" "+self.token+", noop = "+str(r) )
		return r

	def	listDir(self, path, pageSize=10000, cookie=0, stat=True ):
		r = self.api.listDir( self.token, path, pageSize, cookie, stat )
		logger.info( self.uid+" "+self.token+", listDir "+path+" pageSize "+str(pageSize)+" cookie "+str(cookie)+" stat "+str(stat) )
		# +" = "+str(r) )
		return r

	def	listFile(self, path, pageSize=10000, cookie=0, stat=True ):
		r = self.api.listFile( self.token, path, pageSize, cookie, stat )
		logger.info( self.uid+" "+self.token+", listFile "+path+" pageSize "+str(pageSize)+" cookie "+str(cookie)+" stat "+str(stat) )
		# +" = "+str(r) )
		return r

	def	createMultipart(self, path):
		r = self.api.createMultipart( self.token, path)
		logger.info( self.uid+" "+self.token+", createMultipart "+path )
		return r

	def	createMultipartPiece(self, mpid='', number=0, size=0, checksum='', uri=''):
		r = self.api.createMultipartPiece( self.token, mpid, number, size, checksum, uri )
		logger.info( self.uid+" "+self.token+", createMultipartPiece mpid " + mpid + " number " + str(number) + " size " + str(size) + " checksum " + checksum + " uri " + uri )

	def	completeMultipart( self, mpid='' ):
		r = self.api.completeMultipart( self.token, mpid )
		logger.info( self.uid+" "+self.token+", completeMultipart " + mpid )
		return r

	def	getMultipartStatus( self, mpid='' ):
		r = self.api.getMultipartStatus( self.token, mpid )
		logger.info( self.uid+" "+self.token+", getMultipartStatus " + mpid )
		return r

	def	listMultipartPiece( self, mpid='', lastpiece=0, pagesize=100 ):
		r = self.api.listMultipartPiece( self.token, mpid, lastpiece, pagesize )
		logger.info( self.uid+" "+self.token+", listMultipartPiece " + mpid + " lastpiece " +str(lastpiece)+ " pagesize "+ str(pagesize) )
		return r

	def	makeDir(self, path):
		r = self.api.makeDir( self.token, path)
		logger.info( self.uid+" "+self.token+", makeDir "+path+" = "+str(r) )
		return r

	def	makeDir2(self, path):
		r = self.api.makeDir2( self.token, path)
		logger.info( self.uid+" "+self.token+", makeDir2 "+path+" = "+str(r) )
		return r

	def	deleteFile(self, path):
		r = self.api.deleteFile( self.token, path)
		logger.info( self.uid+" "+self.token+", deleteFile "+path+" = "+str(r) )
		return r

	def	rm(self, path):
		if (self.fexists(path)):
			r = self.deleteFile(path)
			if (r == 0): 
				logger.info( self.uid+" "+self.token+", rm "+path+" succeeded" )
				return True
			else: 
				logger.warning( self.uid+" "+self.token+", rm "+path+" failed" )
				return False
		else:
			logger.warning( self.uid+" "+self.token+", rm "+path+" skipped nonexistent file" )

	def	deleteDir(self, path):
		r = self.api.deleteDir( self.token, path)
		logger.info( self.uid+" "+self.token+", deleteDir "+path+" = " + str(r) )
		return r

	def	deleteObject(self, path):
		r = self.api.deleteObject( self.token, path)
		logger.info( self.uid+" "+self.token+", deleteObject "+path+" = " + str(r) )
		return r

	def	rename(self, path, newpath):
		r = self.api.rename( self.token, path, newpath)
		logger.info( self.uid+" "+self.token+", rename "+path+" to "+newpath+" = " + str(r) )
		return r
	
	def	copyFile(self, path, newpath):
		r = self.api.copyFile( self.token, path, newpath)
		logger.info( self.uid+" "+self.token+", copyFile "+path+" to "+newpath+" = " + str(r) )
		return r

	def	registerCallback( self, uri, flags=0, threshold=0 ):
		r = self.api.registerCallback( self.token, uri, flags, threshold )
		logger.info( self.uid+" "+self.token+", registerCallback "+uri+" flags "+str(flags)+" threshold "+str(threshold)+" = " + str(r) )
		return r

	def	listCallback( self ):
		r = self.api.listCallback( self.token )
		logger.info( self.uid+" "+self.token+", listCallback" )
		return r

	def	fetchFileHTTP( self, path, uri, username=None, password=None, auth=None, callbackid=0, priority=0, flags=0, expose_egress='COMPLETE'):
		r = self.api.fetchFileHTTP( self.token, path, uri, username, password, auth, callbackid, priority, flags, expose_egress)
		logger.info( self.uid+" "+self.token+", fetchFileHTTP path "+path+" uri "+uri+" username "+str(username)+" auth "+str(auth)+" callbackid "+str(callbackid)+" priority "+str(priority)+" flags "+str(flags)+" expose_egress "+expose_egress+" = " + str(r) )
		return r

	def	fetchFileFTP(self, path, hostname, filename, username=None, password=None, port=21, passive=True, callbackid=0, priority=0, flags=0, expose_egress='COMPLETE'):
		r = self.api.fetchFileFTP( self.token, path, hostname, filename, username, password, port, passive, callbackid, priority, flags, expose_egress)
		logger.info( self.uid+" "+self.token+", fetchFileFTP path "+path+" hostname "+hostname+" filename "+filename+" username "+str(username)+" port "+str(port)+" passive "+str(passive)+" callbackid "+str(callbackid)+" priority "+str(priority)+" flags "+str(flags)+" expose_egress "+expose_egress+" = " + str(r) )
		return r

	def	fexists( self, path ):
		r = self.stat( path )
		logger.info( self.uid+" "+self.token+", fexists "+path+" = "+str(r) )
		if ((r['code'] == 0) and (r['type'] == 2)): return True
		else: return False

	def	dexists( self, path ):
		r = self.stat( path )
		logger.info( self.uid+" "+self.token+", dexists "+path+" = "+str(r) )
		if ((r['code'] == 0) and (r['type'] == 1)): return True
		else: return False

	def	exists(self, path):
		logger.info( self.uid+" "+self.token+", exists "+path )
		# +" = "+str(r) )
		if (self.fexists(path) or self.dexists(path)): return True
		else: return False

	def	mkdir(self, path, recursive = False):
		logger.info( self.uid+" "+self.token+", mkdir "+path )
		# +" = "+str(r) )
		if (not self.dexists(path)):
			if (recursive):
				r = self.makeDir2( path )
			else:
				r = self.makeDir( path )
			if (r == 0): 
				return True
			else: 
				return False
		else:	
			return False

	def	read(self, path):
		logger.info( self.uid+" "+self.token+", read "+self.mapper.url+urllib2.quote(path) )
		if (self.fexists(path)):
			response = urllib2.urlopen( self.mapperurl+urllib2.quote(path) )
			return response.read()
		else:
			return False

	def	pbar_callback(self, param, current, total):
		if (self.pbar==None):
			widgets = [ unicode(self.pbarfname, errors='ignore').encode('utf-8'), ' ', progressbar.FileTransferSpeed(), ' [', progressbar.Bar(), '] ', progressbar.Percentage(), ' ', progressbar.ETA() ]
			self.pbar = progressbar.ProgressBar( widgets=widgets, maxval=total ).start()
		try:
			self.pbar.update(current)
		except AssertionError, e:
			print e
			print "!"

	def	post(self, sourceObject, targetPath, rename=None, mimetype='auto', mtime=None, egress_policy='COMPLETE', mkdir=False, callback=None):
		logger.info( self.uid+" "+self.token+", post "+sourceObject+" "+targetPath+", rename="+str(rename)+", mimetype="+str(mimetype)+", mtime="+str(mtime)+", egress="+str(egress_policy)+", mkdir="+str(mkdir))
		if (not os.path.isfile(sourceObject)): logger.info( "local("+sourceObject+") does not exist") ; return False 
		if (not self.dexists(targetPath)): logger.info( "remote("+targetPath+") does not exist") ; return False

		sourcePath = os.path.dirname(sourceObject) 
		sourceName = os.path.basename(sourceObject)
		self.pbarfname = sourceName
	
		poster.streaminghttp.register_openers()
	
		datagen, headers = poster.encode.multipart_encode( {
			"uploadFile": open( os.path.join(sourcePath,sourceName), "rb"),
			"directory": targetPath,
			"basename": sourceName,
			"expose_egress": egress_policy
			}, cb=callback)

		request = Request(self.posturl, datagen, headers)

		request.add_header("X-Agile-Authorization", self.token)
		request.add_header("X-Content-Type", mimetype )

		success = False ; attempt = 0
		while not success:
			attempt += 1
			try: 
				result = urlopen(request).read() 
				if callback<>None: self.pbar.finish()
				self.pbar = None
				success = True
			except HTTPError, e: 
				if callback<>None: self.pbar.finish()
				print '[!] HTTP Error: ', e.code
				self.pbar = None
				success = False
			except URLError, e: 
				if callback<>None: self.pbar.finish()
				print '[!] URL Error: ', e.reason
				self.pbar = None
				success = False
